rotkehlchen/chain/evm/decoding/kyber/decoder.py

```diff
@@ -0,0 +1,183 @@
+        _maybe_update_events_legacy_contrats(
+            decoded_events=context.decoded_events,
+            sender=sender,
+            source_asset=source_asset,
+            destination_asset=destination_asset,
+            spent_amount=spent_amount,
+            return_amount=return_amount,
+            notify_user=self.notify_user,
+        )
+
+        return DEFAULT_DECODING_OUTPUT
+
+    def _decode_legacy_upgraded_trade(self, context: DecoderContext) -> DecodingOutput:
+        if context.tx_log.topics[0] != KYBER_TRADE_LEGACY:
+            return DEFAULT_DECODING_OUTPUT
+
+        sender, source_asset, destination_asset = self._legacy_contracts_basic_info(context.tx_log)
+        spent_amount_raw = hex_or_bytes_to_int(context.tx_log.data[96:128])
+        return_amount_raw = hex_or_bytes_to_int(context.tx_log.data[128:160])
+        spent_amount = asset_normalized_value(amount=spent_amount_raw, asset=source_asset)
+        return_amount = asset_normalized_value(amount=return_amount_raw, asset=destination_asset)
+        _maybe_update_events_legacy_contrats(
+            decoded_events=context.decoded_events,
+            sender=sender,
+            source_asset=source_asset,
+            destination_asset=destination_asset,
+            spent_amount=spent_amount,
+            return_amount=return_amount,
+            notify_user=self.notify_user,
+        )
+
+        return DEFAULT_DECODING_OUTPUT
+
+    def _decode_aggregator_trade(self, context: DecoderContext) -> DecodingOutput:
+        if context.tx_log.topics[0] != KYBER_AGGREGATOR_TRADE:
+            return DEFAULT_DECODING_OUTPUT
+
+        sender = hex_or_bytes_to_address(context.tx_log.data[:32])
+        receiver = hex_or_bytes_to_address(context.tx_log.data[96:128])
+
+        if self.base.any_tracked([sender, receiver]) is False:
+            return DEFAULT_DECODING_OUTPUT
+
+        source_token_address = hex_or_bytes_to_address(context.tx_log.data[32:64])
+        destination_token_address = hex_or_bytes_to_address(context.tx_log.data[64:96])
+        spent_amount_raw = hex_or_bytes_to_int(context.tx_log.data[128:160])
+        return_amount_raw = hex_or_bytes_to_int(context.tx_log.data[160:192])
+
+        source_asset = self.base.get_or_create_evm_asset(source_token_address)
+        destination_asset = self.base.get_or_create_evm_asset(destination_token_address)
+        spent_amount = asset_normalized_value(amount=spent_amount_raw, asset=source_asset)
+        return_amount = asset_normalized_value(amount=return_amount_raw, asset=destination_asset)
+        _maybe_update_events_legacy_contrats(
+            decoded_events=context.decoded_events,
+            sender=sender,
+            source_asset=source_asset,
+            destination_asset=destination_asset,
+            spent_amount=spent_amount,
+            return_amount=return_amount,
+            notify_user=self.notify_user,
+        )
```

Comment:
if this logic is valid for all then rename it to `_maybe_update_events` and make it a class method